neuronsimulator/nrn
# one setup.py to rule them all

The gist
---------
* 5 extensions: hoc module, 3 Rx3D extensions and music 
* main setup.py already knew about the extensions (except music, covering that as well)
* this PR drops
  * src/nrnpython/setup.py.in -> hoc_mdule
  * rxd/geometry3d/setup.py.in -> RX3D extensions 
  * rxd/geometry3d/CMakeLists.txt -> lots of non-required logic (targets, install scripts..)
  * src/neuronmusic/setup.py.in -> music extension
  * src/neuronmusic/CMakeLists.txt -> lots of unrequired logic copied over 
* two operational modes:
  * wheel building -> setup.py build_ext bdist_wheel
    * same as before, configure all extensions with cmake, build them, collect them
  * cmake build    -> setup.py build_ext build  
    * provide cmake build folder for build_ext, do not run cmake, build them
    * options: Rx3D + Opt level, music,
    * nrnpython version handling
    * mingw handling
    * drop dependency on VisualC for building extensions

How to test RXD with the Windows Installer
----------------------------------------------------------
* Go to the windows installer GH Actions : https://github.com/neuronsimulator/nrn/actions/workflows/windows.yml?query=branch%3Asetup-py-the-one-and-only
* Click on the first build on that page (for the latest commit) and download `nrn-nightly-AMD64.exe`
* Install that, `git` and an official Python from Python.org
* in a powershell : 
  ```pwsh
  git clone https://github.com/neuronsimulator/nrn.git nrntest
  cd nrntest 
  git submodule update --init
  python .\share\lib\python\neuron\rxdtests\run_all.py
  ```
  you should see: 
  ```
  Failed:
  Incomplete:
  ---
  Ran 50 tests:
  Passed:     50
  Failed:     0
  Incomplete: 0
  ```
   
CI_BRANCHES:SPACK_BRANCH=nrn-1spy
